File: taurus-protect-sdk-typescript/src/mappers/action.ts
/**
 * Action mapper functions for converting OpenAPI DTOs to domain models.
 */

import type {
  Action,
  ActionAmount,
  ActionAttribute,
  ActionComparator,
  ActionDestination,
  ActionEnvelope,
  ActionSource,
  ActionTarget,
  ActionTask,
  ActionTrail,
  ActionTrigger,
  TargetAddress,
  TargetWallet,
  TaskNotification,
  TaskTransfer,
  TriggerBalance,
} from '../models/action';
import { safeBool, safeDate, safeMap, safeString } from './base';

/**
 * Maps a target address DTO to a TargetAddress domain model.
 */
export function targetAddressFromDto(dto: unknown): TargetAddress | undefined {
  if (!dto || typeof dto !== 'object') {
    return undefined;
  }

  const d = dto as Record<string, unknown>;
  return {
    kind: safeString(d.kind),
    addressId: safeString(d.addressID ?? d.addressId ?? d.address_id),
  };
}

/**
 * Maps a target wallet DTO to a TargetWallet domain model.
 */
export function targetWalletFromDto(dto: unknown): TargetWallet | undefined {
  if (!dto || typeof dto !== 'object') {
    return undefined;
  }

  const d = dto as Record<string, unknown>;
  return {
    kind: safeString(d.kind),
    walletId: safeString(d.walletID ?? d.walletId ?? d.wallet_id),
  };
}

/**
 * Maps an action target DTO to an ActionTarget domain model.
 */
export function actionTargetFromDto(dto: unknown): ActionTarget | undefined {
  if (!dto || typeof dto !== 'object') {
    return undefined;
  }

  const d = dto as Record<string, unknown>;
  return {
    kind: safeString(d.kind),
    address: targetAddressFromDto(d.address),
    wallet: targetWalletFromDto(d.wallet),
  };
}

/**
 * Maps an action comparator DTO to an ActionComparator domain model.
 */
export function actionComparatorFromDto(dto: unknown): ActionComparator | undefined {
  if (!dto || typeof dto !== 'object') {
    return undefined;
  }

  const d = dto as Record<string, unknown>;
  return {
    kind: safeString(d.kind),
  };
}

/**
 * Maps an action amount DTO to an ActionAmount domain model.
 */
export function actionAmountFromDto(dto: unknown): ActionAmount | undefined {
  if (!dto || typeof dto !== 'object') {
    return undefined;
  }

  const d = dto as Record<string, unknown>;
  return {
    kind: safeString(d.kind),
    cryptoAmount: safeString(d.cryptoAmount ?? d.crypto_amount),
  };
}

/**
 * Maps a trigger balance DTO to a TriggerBalance domain model.
 */
export function triggerBalanceFromDto(dto: unknown): TriggerBalance | undefined {
  if (!dto || typeof dto !== 'object') {
    return undefined;
  }

  const d = dto as Record<string, unknown>;
  return {
    target: actionTargetFromDto(d.target),
    comparator: actionComparatorFromDto(d.comparator),
    amount: actionAmountFromDto(d.amount),
  };
}

/**
 * Maps an action trigger DTO to an ActionTrigger domain model.
 */
export function actionTriggerFromDto(dto: unknown): ActionTrigger | undefined {
  if (!dto || typeof dto !== 'object') {
    return undefined;
  }

  const d = dto as Record<string, unknown>;
  return {
    kind: safeString(d.kind),
    balance: triggerBalanceFromDto(d.balance),
  };
}

/**
 * Maps an action source DTO to an ActionSource domain model.
 */
export function actionSourceFromDto(dto: unknown): ActionSource | undefined {
  if (!dto || typeof dto !== 'object') {
    return undefined;
  }

  const d = dto as Record<string, unknown>;
  return {
    kind: safeString(d.kind),
    addressId: safeString(d.addressID ?? d.addressId ?? d.address_id),
    walletId: safeString(d.walletID ?? d.walletId ?? d.wallet_id),
  };
}

/**
 * Maps an action destination DTO to an ActionDestination domain model.
 */
export function actionDestinationFromDto(dto: unknown): ActionDestination | undefined {
  if (!dto || typeof dto !== 'object') {
    return undefined;
  }

  const d = dto as Record<string, unknown>;
  return {
    kind: safeString(d.kind),
    addressId: safeString(d.addressID ?? d.addressId ?? d.address_id),
    whitelistedAddressId: safeString(
      d.whitelistedAddressID ?? d.whitelistedAddressId ?? d.whitelisted_address_id
    ),
    walletId: safeString(d.walletID ?? d.walletId ?? d.wallet_id),
  };
}

/**
 * Maps a task transfer DTO to a TaskTransfer domain model.
 */
export function taskTransferFromDto(dto: unknown): TaskTransfer | undefined {
  if (!dto || typeof dto !== 'object') {
    return undefined;
  }

  const d = dto as Record<string, unknown>;
  return {
    from: actionSourceFromDto(d.from),
    to: actionDestinationFromDto(d.to),
    amount: actionAmountFromDto(d.amount),
    topUp: safeBool(d.topUp ?? d.top_up),
    useAllFunds: safeBool(d.useAllFunds ?? d.use_all_funds),
  };
}

/**
 * Maps a task notification DTO to a TaskNotification domain model.
 */
export function taskNotificationFromDto(dto: unknown): TaskNotification | undefined {
  if (!dto || typeof dto !== 'object') {
    return undefined;
  }

  const d = dto as Record<string, unknown>;
  const emailAddresses = d.emailAddresses ?? d.email_addresses;
  return {
    emailAddresses: Array.isArray(emailAddresses)
      ? emailAddresses.map((e) => String(e))
      : undefined,
    notificationMessage: safeString(d.notificationMessage ?? d.notification_message),
    numberOfReminders: safeString(d.numberOfReminders ?? d.number_of_reminders),
  };
}

/**
 * Maps an action task DTO to an ActionTask domain model.
 */
export function actionTaskFromDto(dto: unknown): ActionTask | undefined {
  if (!dto || typeof dto !== 'object') {
    return undefined;
  }

  const d = dto as Record<string, unknown>;
  return {
    kind: safeString(d.kind),
    transfer: taskTransferFromDto(d.transfer),
    notification: taskNotificationFromDto(d.notification),
  };
}

/**
 * Maps an action DTO to an Action domain model.
 */
export function actionFromDto(dto: unknown): Action | undefined {
  if (!dto || typeof dto !== 'object') {
    return undefined;
  }

  const d = dto as Record<string, unknown>;
  const tasks = d.tasks;
  return {
    trigger: actionTriggerFromDto(d.trigger),
    tasks: Array.isArray(tasks) ? safeMap(tasks, actionTaskFromDto) : undefined,
  };
}

/**
 * Maps an action attribute DTO to an ActionAttribute domain model.
 */
export function actionAttributeFromDto(dto: unknown): ActionAttribute | undefined {
  if (!dto || typeof dto !== 'object') {
    return undefined;
  }

  const d = dto as Record<string, unknown>;
  return {
    id: safeString(d.id),
    tenantId: safeString(d.tenantId ?? d.tenant_id),
    key: safeString(d.key),
    value: safeString(d.value),
    contentType: safeString(d.contentType ?? d.content_type),
  };
}

/**
 * Maps an action trail DTO to an ActionTrail domain model.
 */
export function actionTrailFromDto(dto: unknown): ActionTrail | undefined {
  if (!dto || typeof dto !== 'object') {
    return undefined;
  }

  const d = dto as Record<string, unknown>;
  return {
    id: safeString(d.id),
    action: safeString(d.action),
    comment: safeString(d.comment),
    date: safeDate(d.date),
    actionStatus: safeString(d.actionStatus ?? d.action_status),
  };
}

/**
 * Maps an action envelope DTO to an ActionEnvelope domain model.
 */
export function actionEnvelopeFromDto(dto: unknown): ActionEnvelope | undefined {
  if (!dto || typeof dto !== 'object') {
    return undefined;
  }

  const d = dto as Record<string, unknown>;
  const attributes = d.attributes;
  const trails = d.trails;

  return {
    id: safeString(d.id),
    tenantId: safeString(d.tenantId ?? d.tenant_id),
    label: safeString(d.label),
    action: actionFromDto(d.action),
    status: safeString(d.status),
    creationDate: safeDate(d.creationDate ?? d.creation_date),
    updateDate: safeDate(d.updateDate ?? d.update_date),
    lastCheckedDate: safeDate(d.lastcheckeddate ?? d.lastCheckedDate ?? d.last_checked_date),
    autoApprove: safeBool(d.autoApprove ?? d.auto_approve),
    attributes: Array.isArray(attributes)
      ? safeMap(attributes, actionAttributeFromDto)
      : undefined,
    trails: Array.isArray(trails) ? safeMap(trails, actionTrailFromDto) : undefined,
  };
}

/**
 * Maps an array of action envelope DTOs to ActionEnvelope domain models.
 */
export function actionEnvelopesFromDto(
  dtos: unknown[] | null | undefined
): ActionEnvelope[] {
  return safeMap(dtos, actionEnvelopeFromDto);
}

File: taurus-protect-sdk-go/pkg/protect/mapper/action.go
package mapper

import (
	"github.com/taurushq-io/taurus-protect-sdk/taurus-protect-sdk-go/internal/openapi"
	"github.com/taurushq-io/taurus-protect-sdk/taurus-protect-sdk-go/pkg/protect/model"
)

// ActionFromDTO converts an OpenAPI ActionEnvelope to a domain Action.
func ActionFromDTO(dto *openapi.TgvalidatordActionEnvelope) *model.Action {
	if dto == nil {
		return nil
	}

	action := &model.Action{
		ID:          safeString(dto.Id),
		TenantID:    safeString(dto.TenantId),
		Label:       safeString(dto.Label),
		Status:      safeString(dto.Status),
		AutoApprove: safeBool(dto.AutoApprove),
	}

	// Convert dates
	if dto.CreationDate != nil {
		action.CreationDate = *dto.CreationDate
	}
	if dto.UpdateDate != nil {
		action.UpdateDate = *dto.UpdateDate
	}
	if dto.Lastcheckeddate != nil {
		action.LastCheckedDate = *dto.Lastcheckeddate
	}

	// Convert action details
	if dto.Action != nil {
		action.ActionDetails = ActionDetailsFromDTO(dto.Action)
	}

	// Convert attributes
	action.Attributes = ActionAttributesFromDTO(dto.Attributes)

	// Convert trails
	action.Trails = ActionTrailsFromDTO(dto.Trails)

	return action
}

// ActionsFromDTO converts a slice of OpenAPI ActionEnvelope to domain Actions.
func ActionsFromDTO(dtos []openapi.TgvalidatordActionEnvelope) []*model.Action {
	if dtos == nil {
		return nil
	}
	actions := make([]*model.Action, len(dtos))
	for i := range dtos {
		actions[i] = ActionFromDTO(&dtos[i])
	}
	return actions
}

// ActionDetailsFromDTO converts an OpenAPI Action to domain ActionDetails.
func ActionDetailsFromDTO(dto *openapi.TgvalidatordAction) *model.ActionDetails {
	if dto == nil {
		return nil
	}

	details := &model.ActionDetails{}

	if dto.Trigger != nil {
		details.Trigger = ActionTriggerFromDTO(dto.Trigger)
	}

	details.Tasks = ActionTasksFromDTO(dto.Tasks)

	return details
}

// ActionTriggerFromDTO converts an OpenAPI ActionTrigger to domain ActionTrigger.
func ActionTriggerFromDTO(dto *openapi.ActionTrigger) *model.ActionTrigger {
	if dto == nil {
		return nil
	}

	trigger := &model.ActionTrigger{
		Kind: safeString(dto.Kind),
	}

	if dto.Balance != nil {
		trigger.Balance = TriggerBalanceFromDTO(dto.Balance)
	}

	return trigger
}

// TriggerBalanceFromDTO converts an OpenAPI TriggerBalance to domain TriggerBalance.
func TriggerBalanceFromDTO(dto *openapi.TriggerBalance) *model.TriggerBalance {
	if dto == nil {
		return nil
	}

	balance := &model.TriggerBalance{}

	if dto.Target != nil {
		balance.Target = ActionTargetFromDTO(dto.Target)
	}
	if dto.Comparator != nil {
		balance.Comparator = ActionComparatorFromDTO(dto.Comparator)
	}
	if dto.Amount != nil {
		balance.Amount = ActionAmountFromDTO(dto.Amount)
	}

	return balance
}

// ActionTargetFromDTO converts an OpenAPI ActionTarget to domain ActionTarget.
func ActionTargetFromDTO(dto *openapi.ActionTarget) *model.ActionTarget {
	if dto == nil {
		return nil
	}

	target := &model.ActionTarget{
		Kind: safeString(dto.Kind),
	}

	if dto.Address != nil {
		target.Address = TargetAddressFromDTO(dto.Address)
	}
	if dto.Wallet != nil {
		target.Wallet = TargetWalletFromDTO(dto.Wallet)
	}

	return target
}

// TargetAddressFromDTO converts an OpenAPI TargetAddress to domain TargetAddress.
func TargetAddressFromDTO(dto *openapi.TargetAddress) *model.TargetAddress {
	if dto == nil {
		return nil
	}

	return &model.TargetAddress{
		Kind:      safeString(dto.Kind),
		AddressID: safeString(dto.AddressID),
	}
}

// TargetWalletFromDTO converts an OpenAPI TargetWallet to domain TargetWallet.
func TargetWalletFromDTO(dto *openapi.TargetWallet) *model.TargetWallet {
	if dto == nil {
		return nil
	}

	return &model.TargetWallet{
		Kind:     safeString(dto.Kind),
		WalletID: safeString(dto.WalletID),
	}
}

// ActionComparatorFromDTO converts an OpenAPI ActionComparator to domain ActionComparator.
func ActionComparatorFromDTO(dto *openapi.ActionComparator) *model.ActionComparator {
	if dto == nil {
		return nil
	}

	return &model.ActionComparator{
		Kind: safeString(dto.Kind),
	}
}

// ActionAmountFromDTO converts an OpenAPI ActionAmount to domain ActionAmount.
func ActionAmountFromDTO(dto *openapi.TgvalidatordActionAmount) *model.ActionAmount {
	if dto == nil {
		return nil
	}

	return &model.ActionAmount{
		Kind:         safeString(dto.Kind),
		CryptoAmount: safeString(dto.CryptoAmount),
	}
}

// ActionTasksFromDTO converts a slice of OpenAPI ActionTask to domain ActionTasks.
func ActionTasksFromDTO(dtos []openapi.ActionTask) []*model.ActionTask {
	if dtos == nil {
		return nil
	}
	tasks := make([]*model.ActionTask, len(dtos))
	for i := range dtos {
		tasks[i] = ActionTaskFromDTO(&dtos[i])
	}
	return tasks
}

// ActionTaskFromDTO converts an OpenAPI ActionTask to domain ActionTask.
func ActionTaskFromDTO(dto *openapi.ActionTask) *model.ActionTask {
	if dto == nil {
		return nil
	}

	task := &model.ActionTask{
		Kind: safeString(dto.Kind),
	}

	if dto.Transfer != nil {
		task.Transfer = TaskTransferFromDTO(dto.Transfer)
	}
	if dto.Notification != nil {
		task.Notification = TaskNotificationFromDTO(dto.Notification)
	}

	return task
}

// TaskTransferFromDTO converts an OpenAPI TaskTransfer to domain TaskTransfer.
func TaskTransferFromDTO(dto *openapi.TaskTransfer) *model.TaskTransfer {
	if dto == nil {
		return nil
	}

	transfer := &model.TaskTransfer{
		TopUp:       safeBool(dto.TopUp),
		UseAllFunds: safeBool(dto.UseAllFunds),
	}

	if dto.From != nil {
		transfer.From = ActionSourceFromDTO(dto.From)
	}
	if dto.To != nil {
		transfer.To = ActionDestinationFromDTO(dto.To)
	}
	if dto.Amount != nil {
		transfer.Amount = ActionAmountFromDTO(dto.Amount)
	}

	return transfer
}

// ActionSourceFromDTO converts an OpenAPI ActionSource to domain ActionSource.
func ActionSourceFromDTO(dto *openapi.TgvalidatordActionSource) *model.ActionSource {
	if dto == nil {
		return nil
	}

	return &model.ActionSource{
		Kind:      safeString(dto.Kind),
		AddressID: safeString(dto.AddressID),
		WalletID:  safeString(dto.WalletID),
	}
}

// ActionDestinationFromDTO converts an OpenAPI ActionDestination to domain ActionDestination.
func ActionDestinationFromDTO(dto *openapi.TgvalidatordActionDestination) *model.ActionDestination {
	if dto == nil {
		return nil
	}

	return &model.ActionDestination{
		Kind:                 safeString(dto.Kind),
		AddressID:            safeString(dto.AddressID),
		WhitelistedAddressID: safeString(dto.WhitelistedAddressID),
		WalletID:             safeString(dto.WalletID),
	}
}

// TaskNotificationFromDTO converts an OpenAPI TaskNotification to domain TaskNotification.
func TaskNotificationFromDTO(dto *openapi.TaskNotification) *model.TaskNotification {
	if dto == nil {
		return nil
	}

	return &model.TaskNotification{
		EmailAddresses:      dto.EmailAddresses,
		NotificationMessage: safeString(dto.NotificationMessage),
		NumberOfReminders:   safeString(dto.NumberOfReminders),
	}
}

// ActionAttributesFromDTO converts a slice of OpenAPI ActionAttribute to domain ActionAttributes.
func ActionAttributesFromDTO(dtos []openapi.TgvalidatordActionAttribute) []*model.ActionAttribute {
	if dtos == nil {
		return nil
	}
	attrs := make([]*model.ActionAttribute, len(dtos))
	for i := range dtos {
		attrs[i] = ActionAttributeFromDTO(&dtos[i])
	}
	return attrs
}

// ActionAttributeFromDTO converts an OpenAPI ActionAttribute to domain ActionAttribute.
func ActionAttributeFromDTO(dto *openapi.TgvalidatordActionAttribute) *model.ActionAttribute {
	if dto == nil {
		return nil
	}

	return &model.ActionAttribute{
		ID:          safeString(dto.Id),
		TenantID:    safeString(dto.TenantId),
		Key:         safeString(dto.Key),
		Value:       safeString(dto.Value),
		ContentType: safeString(dto.ContentType),
	}
}

// ActionTrailsFromDTO converts a slice of OpenAPI ActionEnvelopeTrail to domain ActionTrails.
func ActionTrailsFromDTO(dtos []openapi.TgvalidatordActionEnvelopeTrail) []*model.ActionTrail {
	if dtos == nil {
		return nil
	}
	trails := make([]*model.ActionTrail, len(dtos))
	for i := range dtos {
		trails[i] = ActionTrailFromDTO(&dtos[i])
	}
	return trails
}

// ActionTrailFromDTO converts an OpenAPI ActionEnvelopeTrail to domain ActionTrail.
func ActionTrailFromDTO(dto *openapi.TgvalidatordActionEnvelopeTrail) *model.ActionTrail {
	if dto == nil {
		return nil
	}

	trail := &model.ActionTrail{
		ID:           safeString(dto.Id),
		Action:       safeString(dto.Action),
		Comment:      safeString(dto.Comment),
		ActionStatus: safeString(dto.ActionStatus),
	}

	if dto.Date != nil {
		trail.Date = *dto.Date
	}

	return trail
}
